Add a funclib util for node_affinity

shashank855 opened the pull request:
**What this PR does / why we need it**:
- Adds a funclib util for node-affinity at path ```funclib/kubectl/pod_node_affinity.yml```.
- Add a litmus book which takes **namespace, application-label, pvc-name** to perform node affinity test.
- The litmus book uses funclib to acheive its logic by passing namespace & pvc-rep-pod name to funclib.

**The following is the log for the litmus job:**
```
PLAY [localhost] ***************************************************************

TASK [Gathering Facts] *********************************************************
ok: [127.0.0.1]

TASK [Generate the litmus result CR to reflect SOT (Start of Test)] ************
changed: [127.0.0.1] => {"changed": true, "checksum": "d89a8832dae48108e8003ae6344c4bf38902540a", "dest": "./litmus-result.yaml", "gid": 0, "group": "root", "md5sum": "44902e05bd5c90b18e615a22d6a31ccc", "mode": "0644", "owner": "root", "size": 414, "src": "/root/.ansible/tmp/ansible-tmp-1538725649.35-281038777516226/source", "state": "file", "uid": 0}

TASK [Apply the litmus result CR] **********************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl apply -f litmus-result.yaml", "delta": "0:00:01.659930", "end": "2018-10-05 07:47:33.492287", "failed_when_result": false, "rc": 0, "start": "2018-10-05 07:47:31.832357", "stderr": "", "stderr_lines": [], "stdout": "litmusresult \"node-affinity\" configured", "stdout_lines": ["litmusresult \"node-affinity\" configured"]}

TASK [Getting Pod name of Application] *****************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl get pod -n litmus -l lkey=lvalue -o jsonpath='{.items[0].metadata.name}'", "delta": "0:00:01.319745", "end": "2018-10-05 07:47:35.466994", "rc": 0, "start": "2018-10-05 07:47:34.147249", "stderr": "", "stderr_lines": [], "stdout": "pgset-0", "stdout_lines": ["pgset-0"]}

TASK [Getting PVC name] ********************************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl get pod -n litmus -l pvc=crunchy-claim-pgset-0 | awk 'NR == 3 {print $1}'", "delta": "0:00:01.305643", "end": "2018-10-05 07:47:37.365334", "rc": 0, "start": "2018-10-05 07:47:36.059691", "stderr": "", "stderr_lines": [], "stdout": "pvc-8efc32a2-c7b7-11e8-badd-0cc47a2059bc-rep-6ff65c49b8-7tmdw", "stdout_lines": ["pvc-8efc32a2-c7b7-11e8-badd-0cc47a2059bc-rep-6ff65c49b8-7tmdw"]}

TASK [Checking the status of replia-PVC pod] ***********************************
changed: [127.0.0.1] => {"attempts": 1, "changed": true, "cmd": "kubectl get pod pvc-8efc32a2-c7b7-11e8-badd-0cc47a2059bc-rep-6ff65c49b8-7tmdw -n litmus -o jsonpath='{.status.phase}'", "delta": "0:00:01.313909", "end": "2018-10-05 07:47:39.267137", "rc": 0, "start": "2018-10-05 07:47:37.953228", "stderr": "", "stderr_lines": [], "stdout": "Running", "stdout_lines": ["Running"]}

TASK [Checking the node affinity for rep-PVC pod] ******************************
included: /funclib/kubectl/pod_node_affinity.yml for 127.0.0.1

TASK [Getting the Node name for the given pod] *********************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl get pod pvc-8efc32a2-c7b7-11e8-badd-0cc47a2059bc-rep-6ff65c49b8-7tmdw -n litmus -o jsonpath='{.spec.nodeName}'", "delta": "0:00:01.318038", "end": "2018-10-05 07:47:41.465804", "rc": 0, "start": "2018-10-05 07:47:40.147766", "stderr": "", "stderr_lines": [], "stdout": "node-elegant-varahamihira02", "stdout_lines": ["node-elegant-varahamihira02"]}

TASK [Cordening the Node] ******************************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl cordon node-elegant-varahamihira02", "delta": "0:00:01.217484", "end": "2018-10-05 07:47:43.247085", "rc": 0, "start": "2018-10-05 07:47:42.029601", "stderr": "", "stderr_lines": [], "stdout": "node \"node-elegant-varahamihira02\" cordoned", "stdout_lines": ["node \"node-elegant-varahamihira02\" cordoned"]}

TASK [Checking whether Node is cordened] ***************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl get nodes node-elegant-varahamihira02 -o jsonpath='{.spec.unschedulable}'", "delta": "0:00:01.216251", "end": "2018-10-05 07:47:45.023566", "failed_when_result": false, "rc": 0, "start": "2018-10-05 07:47:43.807315", "stderr": "", "stderr_lines": [], "stdout": "true", "stdout_lines": ["true"]}

TASK [Deleting the specified pod] **********************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl delete pod pvc-8efc32a2-c7b7-11e8-badd-0cc47a2059bc-rep-6ff65c49b8-7tmdw -n litmus", "delta": "0:00:01.346631", "end": "2018-10-05 07:47:46.978780", "rc": 0, "start": "2018-10-05 07:47:45.632149", "stderr": "", "stderr_lines": [], "stdout": "pod \"pvc-8efc32a2-c7b7-11e8-badd-0cc47a2059bc-rep-6ff65c49b8-7tmdw\" deleted", "stdout_lines": ["pod \"pvc-8efc32a2-c7b7-11e8-badd-0cc47a2059bc-rep-6ff65c49b8-7tmdw\" deleted"]}

TASK [Getting the Pod name in Pending state] ***********************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl get pods -n litmus -o jsonpath='{.items[?(@.status.phase==\"Pending\")].metadata.name}'", "delta": "0:00:01.346085", "end": "2018-10-05 07:47:48.933470", "rc": 0, "start": "2018-10-05 07:47:47.587385", "stderr": "", "stderr_lines": [], "stdout": "pvc-8efc32a2-c7b7-11e8-badd-0cc47a2059bc-rep-6ff65c49b8-r5gks", "stdout_lines": ["pvc-8efc32a2-c7b7-11e8-badd-0cc47a2059bc-rep-6ff65c49b8-r5gks"]}

TASK [Uncordoning Node] ********************************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl uncordon node-elegant-varahamihira02", "delta": "0:00:01.214588", "end": "2018-10-05 07:47:50.716542", "rc": 0, "start": "2018-10-05 07:47:49.501954", "stderr": "", "stderr_lines": [], "stdout": "node \"node-elegant-varahamihira02\" uncordoned", "stdout_lines": ["node \"node-elegant-varahamihira02\" uncordoned"]}

TASK [Checking for Running status of pod] **************************************
FAILED - RETRYING: Checking for Running status of pod (5 retries left).
changed: [127.0.0.1] => {"attempts": 2, "changed": true, "cmd": "kubectl get pod pvc-8efc32a2-c7b7-11e8-badd-0cc47a2059bc-rep-6ff65c49b8-r5gks -n litmus -o jsonpath='{.status.phase}'", "delta": "0:00:01.342278", "end": "2018-10-05 07:48:04.314606", "rc": 0, "start": "2018-10-05 07:48:02.972328", "stderr": "", "stderr_lines": [], "stdout": "Running", "stdout_lines": ["Running"]}

TASK [Checking whether POD is scheduled on same NODE] **************************
 [WARNING]: when statements should not include jinja2 templating delimiters
such as {{ }} or {% %}. Found: '{{ node_name.stdout }}' not in
node_schedule.stdout
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl get pod pvc-8efc32a2-c7b7-11e8-badd-0cc47a2059bc-rep-6ff65c49b8-r5gks -n litmus -o jsonpath='{.spec.nodeName}'", "delta": "0:00:01.325700", "end": "2018-10-05 07:48:06.246569", "failed_when_result": false, "rc": 0, "start": "2018-10-05 07:48:04.920869", "stderr": "", "stderr_lines": [], "stdout": "node-elegant-varahamihira02", "stdout_lines": ["node-elegant-varahamihira02"]}

TASK [Checking the status of replia-PVC pod] ***********************************
changed: [127.0.0.1] => {"attempts": 1, "changed": true, "cmd": "kubectl get po pvc-8efc32a2-c7b7-11e8-badd-0cc47a2059bc-rep-6ff65c49b8-r5gks -n litmus -o jsonpath='{.status.phase}'", "delta": "0:00:01.297199", "end": "2018-10-05 07:48:08.149735", "rc": 0, "start": "2018-10-05 07:48:06.852536", "stderr": "", "stderr_lines": [], "stdout": "Running", "stdout_lines": ["Running"]}

TASK [set_fact] ****************************************************************
ok: [127.0.0.1] => {"ansible_facts": {"flag": "Pass"}, "changed": false}

TASK [Generate the litmus result CR to reflect EOT (End of Test)] **************
changed: [127.0.0.1] => {"changed": true, "checksum": "f81270ba93f20897ea713a443397bd63ef09c1e9", "dest": "./litmus-result.yaml", "gid": 0, "group": "root", "md5sum": "f9f7fc124c96b670e2b916409bf0ad7a", "mode": "0644", "owner": "root", "size": 412, "src": "/root/.ansible/tmp/ansible-tmp-1538725688.62-91168826805826/source", "state": "file", "uid": 0}

TASK [Apply the litmus result CR] **********************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl apply -f litmus-result.yaml", "delta": "0:00:01.395094", "end": "2018-10-05 07:48:12.938355", "failed_when_result": false, "rc": 0, "start": "2018-10-05 07:48:11.543261", "stderr": "", "stderr_lines": [], "stdout": "litmusresult \"node-affinity\" configured", "stdout_lines": ["litmusresult \"node-affinity\" configured"]}

PLAY RECAP *********************************************************************
127.0.0.1                  : ok=19   changed=16   unreachable=0    failed=0
```

gprasath commented on funclib/kubectl/pod_node_affinity.yml:
```diff
@@ -0,0 +1,40 @@
+  failed_when: "'true' not in result.stdout" 
+
+- name: Deleting the specified pod
+  shell: kubectl delete pod {{ pod_name }} -n {{ ns }}
```
Does the application pod has node affinity specs?

gprasath commented on apps/crunchy-postgres/tests/node_affinity/run_litmus_test.yml:
```diff
@@ -0,0 +1,37 @@
+apiVersion: batch/v1
+kind: Job
+metadata:
+  generateName: litmus
```
modify the name relevant to test

gprasath commented on apps/crunchy-postgres/tests/node_affinity/run_litmus_test.yml:
```diff
@@ -0,0 +1,37 @@
+      restartPolicy: Never
+      containers:
+      - name: ansibletest
+        image: ranjnshashank855/ansible-runner:test
```
Change the image

gprasath commented on apps/crunchy-postgres/tests/node_affinity/run_litmus_test.yml:
```diff
@@ -0,0 +1,37 @@
+            value: litmus
+
+          - name: APPLICATION_LABEL
+            value: lkey=lvalue
```
Give the application label here

gprasath commented on apps/crunchy-postgres/tests/node_affinity/test.yml:
```diff
@@ -0,0 +1,81 @@
+          register: pod_name
+          
+        - name: Getting PVC name
+          shell: kubectl get pod -n {{ namespace }} -l pvc={{ pvc_name }}-{{ pod_name.stdout }} | awk 'NR == 3 {print $1}'
```
Let us use json path to derive one of the replicas, instead of awk

gprasath commented on apps/crunchy-postgres/tests/node_affinity/test.yml:
```diff
@@ -0,0 +1,81 @@
+          shell: kubectl get pod -n {{ namespace }} -l {{ app_label }} -o jsonpath='{.items[0].metadata.name}'
+          register: pod_name
+          
+        - name: Getting PVC name
```
Give relevant task name, here you are fetching pod name.

gprasath commented on apps/crunchy-postgres/tests/node_affinity/test.yml:
```diff
@@ -0,0 +1,81 @@
+          shell: kubectl get po {{ new_pod_name.stdout }} -n {{ namespace }} -o jsonpath='{.status.phase}'
+          until: "'Running' in status.stdout"
+          delay: 5
+          retires: 10      
```
typo.

Increase the interval.

nsathyaseelan commented on apps/crunchy-postgres/tests/node_affinity/run_litmus_test.yml:
```diff
@@ -0,0 +1,37 @@
+            value: default
+
+          - name: NAMESPACE
+            value: litmus
```
@shashank855 use APP_NAMESPACE as a name

shashank855 replied:
@nsathyaseelan I have made the changes

nsathyaseelan replied:
@shashank855 the application namespace will be `app-pgres-ns`

shashank855 replied:
@nsathyaseelan Sure !! I will make the changes.

nsathyaseelan commented on apps/crunchy-postgres/tests/node_affinity/run_litmus_test.yml:
```diff
@@ -0,0 +1,37 @@
+            value: app=pgset
+
+          - name: PVC_NAME
+            value: crunchy-claim   
```
@shashank855 use APP_PVC instead off PVC_NAME

shashank855 replied:
sure, I will make the chenges.

nsathyaseelan commented on apps/crunchy-postgres/tests/node_affinity/test_vars.yml:
```diff
@@ -0,0 +1,4 @@
+test_name: node-affinity
+namespace: "{{ lookup('env','NAMESPACE') }}" 
```
@shashank855  can you change the NAMESPACE in to APP_NAMESPACE?

shashank855 commented:
**Output Achived after using patch file in ```/common/util``` to get node affinity property:**
```
PLAY [localhost] ***************************************************************

TASK [Gathering Facts] *********************************************************
ok: [127.0.0.1]

TASK [Generate the litmus result CR to reflect SOT (Start of Test)] ************
changed: [127.0.0.1] => {"changed": true, "checksum": "d89a8832dae48108e8003ae6344c4bf38902540a", "dest": "./litmus-result.yaml", "gid": 0, "group": "root", "md5sum": "44902e05bd5c90b18e615a22d6a31ccc", "mode": "0644", "owner": "root", "size": 414, "src": "/root/.ansible/tmp/ansible-tmp-1539175447.03-235482092265742/source", "state": "file", "uid": 0}

TASK [Apply the litmus result CR] **********************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl apply -f litmus-result.yaml", "delta": "0:00:01.984661", "end": "2018-10-10 12:44:11.952724", "failed_when_result": false, "rc": 0, "start": "2018-10-10 12:44:09.968063", "stderr": "", "stderr_lines": [], "stdout": "litmusresult.litmus.io/node-affinity configured", "stdout_lines": ["litmusresult.litmus.io/node-affinity configured"]}

TASK [Getting Pod name of Application] *****************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl get pod -n app-pgres-ns -l app=pgset -o jsonpath='{.items[0].metadata.name}'", "delta": "0:00:01.479568", "end": "2018-10-10 12:44:14.115126", "rc": 0, "start": "2018-10-10 12:44:12.635558", "stderr": "", "stderr_lines": [], "stdout": "pgset-0", "stdout_lines": ["pgset-0"]}

TASK [Getting replica-pod name] ************************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl get pod -n app-pgres-ns -l openebs.io/persistent-volume-claim=pgdata-claim-pgset-0 -o jsonpath='{.items[1].metadata.name}'", "delta": "0:00:01.519264", "end": "2018-10-10 12:44:16.292134", "rc": 0, "start": "2018-10-10 12:44:14.772870", "stderr": "", "stderr_lines": [], "stdout": "app-pgres-ns-pgdata-claim-pgset-0-4100344701-rep-77488466fdkjlp", "stdout_lines": ["app-pgres-ns-pgdata-claim-pgset-0-4100344701-rep-77488466fdkjlp"]}

TASK [Checking the status of replia pod] ***************************************
changed: [127.0.0.1] => {"attempts": 1, "changed": true, "cmd": "kubectl get pod app-pgres-ns-pgdata-claim-pgset-0-4100344701-rep-77488466fdkjlp -n app-pgres-ns -o jsonpath='{.status.phase}'", "delta": "0:00:01.470031", "end": "2018-10-10 12:44:18.435499", "rc": 0, "start": "2018-10-10 12:44:16.965468", "stderr": "", "stderr_lines": [], "stdout": "Running", "stdout_lines": ["Running"]}

TASK [Getting the Node name on which replica-pod is scheduled] *****************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl get pod app-pgres-ns-pgdata-claim-pgset-0-4100344701-rep-77488466fdkjlp -n app-pgres-ns -o jsonpath='{.spec.nodeName}'", "delta": "0:00:01.508433", "end": "2018-10-10 12:44:20.613371", "rc": 0, "start": "2018-10-10 12:44:19.104938", "stderr": "", "stderr_lines": [], "stdout": "node-naughty-nobel03", "stdout_lines": ["node-naughty-nobel03"]}

TASK [Obtaing the deployment name for the replica pod] *************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl get deployments. -n app-pgres-ns -o jsonpath='{.items[1].metadata.name}'", "delta": "0:00:01.552080", "end": "2018-10-10 12:44:22.823308", "rc": 0, "start": "2018-10-10 12:44:21.271228", "stderr": "", "stderr_lines": [], "stdout": "app-pgres-ns-pgdata-claim-pgset-0-4100344701-rep", "stdout_lines": ["app-pgres-ns-pgdata-claim-pgset-0-4100344701-rep"]}

TASK [Replacing the Node name in Patch yaml] ***********************************
changed: [127.0.0.1] => {"changed": true, "msg": "1 replacements made"}

TASK [Update the replica deployment with node affinity property] ***************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl patch deployment app-pgres-ns-pgdata-claim-pgset-0-4100344701-rep -n app-pgres-ns -p \"$(cat /common/utils/replica_patch.yml)\"", "delta": "0:00:01.487566", "end": "2018-10-10 12:44:26.204797", "failed_when_result": false, "rc": 0, "start": "2018-10-10 12:44:24.717231", "stderr": "", "stderr_lines": [], "stdout": "deployment.extensions/app-pgres-ns-pgdata-claim-pgset-0-4100344701-rep patched (no change)", "stdout_lines": ["deployment.extensions/app-pgres-ns-pgdata-claim-pgset-0-4100344701-rep patched (no change)"]}

TASK [Checking the node affinity for replica-pod] ******************************
included: /funclib/kubectl/pod_node_affinity.yml for 127.0.0.1

TASK [Getting the Node name for the given pod] *********************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl get pod app-pgres-ns-pgdata-claim-pgset-0-4100344701-rep-77488466fdkjlp -n app-pgres-ns -o jsonpath='{.spec.nodeName}'", "delta": "0:00:01.480348", "end": "2018-10-10 12:44:28.658029", "rc": 0, "start": "2018-10-10 12:44:27.177681", "stderr": "", "stderr_lines": [], "stdout": "node-naughty-nobel03", "stdout_lines": ["node-naughty-nobel03"]}

TASK [Cordening the Node] ******************************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl cordon node-naughty-nobel03", "delta": "0:00:01.482373", "end": "2018-10-10 12:44:30.761957", "rc": 0, "start": "2018-10-10 12:44:29.279584", "stderr": "", "stderr_lines": [], "stdout": "node/node-naughty-nobel03 cordoned", "stdout_lines": ["node/node-naughty-nobel03 cordoned"]}

TASK [Checking whether Node is cordened] ***************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl get nodes node-naughty-nobel03 -o jsonpath='{.spec.unschedulable}'", "delta": "0:00:01.506608", "end": "2018-10-10 12:44:32.886688", "failed_when_result": false, "rc": 0, "start": "2018-10-10 12:44:31.380080", "stderr": "", "stderr_lines": [], "stdout": "true", "stdout_lines": ["true"]}

TASK [Deleting the specified pod] **********************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl delete pod app-pgres-ns-pgdata-claim-pgset-0-4100344701-rep-77488466fdkjlp -n app-pgres-ns", "delta": "0:00:03.342826", "end": "2018-10-10 12:44:36.905374", "rc": 0, "start": "2018-10-10 12:44:33.562548", "stderr": "", "stderr_lines": [], "stdout": "pod \"app-pgres-ns-pgdata-claim-pgset-0-4100344701-rep-77488466fdkjlp\" deleted", "stdout_lines": ["pod \"app-pgres-ns-pgdata-claim-pgset-0-4100344701-rep-77488466fdkjlp\" deleted"]}

TASK [Getting the Pod name in Pending state] ***********************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl get pods -n app-pgres-ns -o jsonpath='{.items[?(@.status.phase==\"Pending\")].metadata.name}'", "delta": "0:00:01.566243", "end": "2018-10-10 12:44:39.124675", "rc": 0, "start": "2018-10-10 12:44:37.558432", "stderr": "", "stderr_lines": [], "stdout": "app-pgres-ns-pgdata-claim-pgset-0-4100344701-rep-77488466fz84nx", "stdout_lines": ["app-pgres-ns-pgdata-claim-pgset-0-4100344701-rep-77488466fz84nx"]}

TASK [Uncordoning Node] ********************************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl uncordon node-naughty-nobel03", "delta": "0:00:01.517887", "end": "2018-10-10 12:44:41.264749", "rc": 0, "start": "2018-10-10 12:44:39.746862", "stderr": "", "stderr_lines": [], "stdout": "node/node-naughty-nobel03 uncordoned", "stdout_lines": ["node/node-naughty-nobel03 uncordoned"]}

TASK [Checking for Running status of pod] **************************************
FAILED - RETRYING: Checking for Running status of pod (5 retries left).
changed: [127.0.0.1] => {"attempts": 2, "changed": true, "cmd": "kubectl get pod app-pgres-ns-pgdata-claim-pgset-0-4100344701-rep-77488466fz84nx -n app-pgres-ns -o jsonpath='{.status.phase}'", "delta": "0:00:01.513249", "end": "2018-10-10 12:44:55.328210", "rc": 0, "start": "2018-10-10 12:44:53.814961", "stderr": "", "stderr_lines": [], "stdout": "Running", "stdout_lines": ["Running"]}

TASK [Checking whether POD is scheduled on same NODE] **************************
 [WARNING]: when statements should not include jinja2 templating delimiters
such as {{ }} or {% %}. Found: '{{ node_name.stdout }}' not in
node_schedule.stdout
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl get pod app-pgres-ns-pgdata-claim-pgset-0-4100344701-rep-77488466fz84nx -n app-pgres-ns -o jsonpath='{.spec.nodeName}'", "delta": "0:00:01.515749", "end": "2018-10-10 12:44:57.519158", "failed_when_result": false, "rc": 0, "start": "2018-10-10 12:44:56.003409", "stderr": "", "stderr_lines": [], "stdout": "node-naughty-nobel03", "stdout_lines": ["node-naughty-nobel03"]}

TASK [Checking the status of replica-pod] **************************************
changed: [127.0.0.1] => {"attempts": 1, "changed": true, "cmd": "kubectl get pod app-pgres-ns-pgdata-claim-pgset-0-4100344701-rep-77488466fz84nx -n app-pgres-ns -o jsonpath='{.status.phase}'", "delta": "0:00:01.554135", "end": "2018-10-10 12:44:59.754172", "rc": 0, "start": "2018-10-10 12:44:58.200037", "stderr": "", "stderr_lines": [], "stdout": "Running", "stdout_lines": ["Running"]}

TASK [set_fact] ****************************************************************
ok: [127.0.0.1] => {"ansible_facts": {"flag": "Pass"}, "changed": false}

TASK [Generate the litmus result CR to reflect EOT (End of Test)] **************
changed: [127.0.0.1] => {"changed": true, "checksum": "f81270ba93f20897ea713a443397bd63ef09c1e9", "dest": "./litmus-result.yaml", "gid": 0, "group": "root", "md5sum": "f9f7fc124c96b670e2b916409bf0ad7a", "mode": "0644", "owner": "root", "size": 412, "src": "/root/.ansible/tmp/ansible-tmp-1539175500.29-268469919623764/source", "state": "file", "uid": 0}

TASK [Apply the litmus result CR] **********************************************
changed: [127.0.0.1] => {"changed": true, "cmd": "kubectl apply -f litmus-result.yaml", "delta": "0:00:01.776048", "end": "2018-10-10 12:45:05.749058", "failed_when_result": false, "rc": 0, "start": "2018-10-10 12:45:03.973010", "stderr": "", "stderr_lines": [], "stdout": "litmusresult.litmus.io/node-affinity configured", "stdout_lines": ["litmusresult.litmus.io/node-affinity configured"]}

PLAY RECAP *********************************************************************
127.0.0.1                  : ok=23   changed=20   unreachable=0    failed=0 
```

gprasath commented on apps/crunchy-postgres/tests/node_affinity/run_litmus_test.yml:
```diff
@@ -0,0 +1,37 @@
+apiVersion: batch/v1
+kind: Job
+metadata:
+  generateName: litmus-node-affinity
```
keep '-' at the end.

gprasath commented on apps/crunchy-postgres/tests/node_affinity/test.yml:
```diff
@@ -0,0 +1,100 @@
+          register: pod_name
+          
+        - name: Getting replica-pod name
+          shell: kubectl get pod -n {{ namespace }} -l openebs.io/persistent-volume-claim={{ pvc_name }}-{{ pod_name.stdout }} -o jsonpath='{.items[1].metadata.name}'
```
label has to be changed according to 0.7.

dargasudarshan replied:
@shashank855 Replica pod does not contain application pod traces in its name, then how are we getting the replica pod name here? Can you please explain?

shashank855 replied:
@dargasudarshan Replica **pod's label** contain the traces the of application pod name. We are getting the replica pod name using this label here.

dargasudarshan replied:
@shashank855 I have checked the same label does not have application pod traces in case of deployment?
Please see the snippet below. Can we make generic to work for both sts and deployments? Just a taught..
```
sudarshan_darga@cloudshell:~/test-tools/dd-client (strong-eon-153112)$ kubectl get pods pvc-f2be77f6-d05c-11e8-881b-42010a80026a-rep-885694bb9-dhgc7 -n app-percona-ns -o yaml
apiVersion: v1
kind: Pod
metadata:
  annotations:
    openebs.io/capacity: 5G
    openebs.io/storage-pool: default
  creationTimestamp: 2018-10-15T12:26:43Z
  generateName: pvc-f2be77f6-d05c-11e8-881b-42010a80026a-rep-885694bb9-
  labels:
    openebs.io/persistent-volume: pvc-f2be77f6-d05c-11e8-881b-42010a80026a
    openebs.io/persistent-volume-claim: percona-mysql-claim
```

nsathyaseelan commented on funclib/kubectl/pod_node_affinity.yml:
```diff
@@ -0,0 +1,40 @@
+  shell: kubectl get pod {{ pod_name }} -n {{ ns }} -o jsonpath='{.spec.nodeName}'
+  register: node_name
+
+- name: Cordening the Node
```
typo corden ->cordon

nsathyaseelan commented on funclib/kubectl/pod_node_affinity.yml:
```diff
@@ -0,0 +1,40 @@
+- name: Cordening the Node
+  shell: kubectl cordon {{ node_name.stdout }}  
+
+- name: Checking whether Node is cordened
```
typo corden ->cordon

gprasath commented on funclib/kubectl/pod_node_affinity.yml:
```diff
@@ -0,0 +1,40 @@
+  shell: kubectl delete pod {{ pod_name }} -n {{ ns }}
+
+- name: Getting the Pod name in Pending state
+  shell: kubectl get pods -n {{ ns }} -o jsonpath='{.items[?(@.status.phase=="Pending")].metadata.name}'
```
include pod label in the condition. Just to eliminate the listing of any other pod in pending state.

gprasath commented on funclib/kubectl/pod_node_affinity.yml:
```diff
@@ -0,0 +1,40 @@
+  register: node_name
+
+- name: Cordoning the Node
+  shell: kubectl cordon {{ node_name.stdout }}  
```
include a condition

gprasath commented on funclib/kubectl/pod_node_affinity.yml:
```diff
@@ -0,0 +1,40 @@
+- name: Checking whether POD is scheduled on same NODE
+  shell: kubectl get pod {{ new_pod_name.stdout }} -n {{ ns }} -o jsonpath='{.spec.nodeName}'
+  register: node_schedule
+  failed_when: "'{{ node_name.stdout }}' not in node_schedule.stdout"      
```
remove jinja template in when condition

gprasath commented on apps/crunchy-postgres/tests/node_affinity/test.yml:
```diff
@@ -0,0 +1,102 @@
+      always:    
+
+        - name: Uncordon node                          #This Task is added inside always because if it the fails node will be available for scheduling.
+         shell: kubectl uncordon {{ node_name.stdout }}
```
indentation
